### ISSUE TITLE
StackOverflowError when compiling templates with multiple recursive partial invocations

### ISSUE THREAD

**Issue by orirawlings** [spullara/mustache.java#126]
We recently came across an issue where compiling one of our mustache templates was causing infinite recursion and a `StackOverflowError`.

We are using version 0.8.15.

After some debugging we realized that the problem occurs when our template includes multiple recursive calls to the same partial. For example:

    {{! In multiple_recursive_partials.html file }}
    {{#type.foo}}
    I'm a foo!
    {{#items}}In foo: {{> multiple_recursive_partials}}{{/items}}
    {{/type.foo}}
    {{#type.bar}}
    I'm a bar!
    {{#items}}In bar: {{> multiple_recursive_partials}}{{/items}}
    {{/type.bar}}

We walked through the code a bit and identified a bug introduced in this commit c9e64e12d0193ffa5fc2b7fa2c7c0ea05067281d

      /**
       * In order to handle recursion, we need a temporary thread local cache during compilation
       * that is ultimately thrown away after the top level partial is complete.
       *
       * @param s
       * @return
       */
      public Mustache compilePartial(String s) {
        Map<String, Mustache> cache = partialCache.get();
        try {
          Mustache mustache = cache.get(s);
          if (mustache == null) {
            mustache = mc.compile(s);
            cache.put(s, mustache);
            mustache.init();
          }
          return mustache;
        } finally {
          cache.remove(s);
        }
      }

Basically, the partials are being removed from the new `partialCache` `Map` earlier than they should be, causing initialization of the second partial invocation to miss this cache and proceed incorrectly down another recursive initialization call.

**Comment by stephenliang** [comment]
+1, any updates on this issue?

**Comment by orirawlings** [comment]
This has been fixed in 0.9.1
